BossExpertise.cs

```diff
@@ -66,38 +64,92 @@ public override void HandlePacket(BinaryReader reader, int whoAmI)
+		void SyncDifficultyMode(int difficulty, int ignoreClient = -1)
+		{ 
+			HookDifficultyMode(difficulty);
+			if (Main.netMode != NetmodeID.SinglePlayer)
 			{
 				var msg = GetPacket();
-				msg.Write((byte)ExpertMessageType.SyncExpert);
-				msg.Write(expert);
+				msg.Write((byte)ExpertMessageType.SyncDifficulty);
+				msg.Write(difficulty);
 				msg.Send(ignoreClient: ignoreClient);
 			}
 		}
 		
-		public void SetExpertMode(bool expert)
+		public static void HookDifficultyMode(int difficulty)
+        {
+			if (difficulty == 0 || difficulty > 2)
+            {
+				FieldInfo expert = typeof(Main)
+				.GetField("_overrideForExpertMode", BindingFlags.Static | BindingFlags.NonPublic);
+				expert.SetValue(null, false);
+				FieldInfo master = typeof(Main)
+				.GetField("_overrideForMasterMode", BindingFlags.Static | BindingFlags.NonPublic);
+				master.SetValue(null, false);
+			}
+			else if (difficulty == 1)
+			{
+				FieldInfo expert = typeof(Main)
+				.GetField("_overrideForExpertMode", BindingFlags.Static | BindingFlags.NonPublic);
+				expert.SetValue(null, true);
+				FieldInfo master = typeof(Main)
+				.GetField("_overrideForMasterMode", BindingFlags.Static | BindingFlags.NonPublic);
+				master.SetValue(null, false);
+
+			}
+			else if (difficulty == 2)
+			{
+				FieldInfo expert = typeof(Main)
+				.GetField("_overrideForExpertMode", BindingFlags.Static | BindingFlags.NonPublic);
+				expert.SetValue(null, true);
+				FieldInfo master = typeof(Main)
+				.GetField("_overrideForMasterMode", BindingFlags.Static | BindingFlags.NonPublic);
+				master.SetValue(null, true);
+			}
+		}
+
+		public void SetDifficultyMode(int difficulty)
 		{
-			if(Main.expertMode && !expert)
+			if (difficulty > 2)
+			{
+				difficulty = 0;
+			}
+			if (Main.expertMode && difficulty == 0)
 			{
-				SyncExpertMode(false);
 				Main.NewText(Language.GetTextValue("Mods.BossExpertise.NowNormalMode"));
 			}
-			else if(!Main.expertMode && expert)
+			else if (!Main.expertMode && difficulty == 1)
 			{
-				SyncExpertMode(true);
 				Main.NewText(Language.GetTextValue("Mods.BossExpertise.NowExpertMode"), 255, 50, 50);
 			}
-			else if(!Main.expertMode && !expert)
+			else if (!Main.masterMode && difficulty == 2)
+			{
+				Main.NewText(Language.GetTextValue("Mods.BossExpertise.NowMasterMode"), 255, 50, 50);
+			}
+			else if (!Main.expertMode && difficulty == 0)
 			{
 				Main.NewText(Language.GetTextValue("Mods.BossExpertise.AlreadyNormalMode"));
 			}
-			else
+			else if (Main.expertMode && difficulty == 1)
 			{
 				Main.NewText(Language.GetTextValue("Mods.BossExpertise.AlreadyExpertMode"), 255, 50, 50);
 			}
+			else if (!Main.masterMode && difficulty == 2)
+			{
+				Main.NewText(Language.GetTextValue("Mods.BossExpertise.AlreadyMasterMode"), 255, 50, 50);
+			}
+			else
+			{
+				Log("Something went wrong in BossExpertise: " + CurrentDifficulty);
+				return;
+			}
+			SyncDifficultyMode(difficulty);
+			CurrentDifficulty = difficulty;
```

Comment:
Here we should change `ActualDifficulty`, not `CurrentDifficulty`, no?

Reply:
Yes, it should have been but that is currently never ran because its part of the chat command. Fixed anyway.